### DISCUSSION TITLE
Deleted files not showing in Github Action when using git log

### DISCUSSION THREAD

**Discussion by JoshMealing** [community/community#78117] [Actions] [answered]
### Select Topic Area

Question

### Body

Hey everyone!

I'm experiencing an issue with one of my github actions. I'll give some background on what I am trying to achieve, and then the issue that I am facing.

**What I'm trying to do:**
Upon a merge into the master branch, I am trying to get all changes and push them to a number of other branches. If a file is deleted in the latest merge commit to master, I would like to remove that file from each of the other branches.

**The problem at hand:**
When trying to get the names of files which have been deleted in the latest merge commit, these are not being output. After running the same command on my local, I am getting the expected output. I'll attach my worker as well as the output that I am expecting from this command.

Any help would be greatly appreciated.

PS: If there is any additional info I should share, please let me know.

Github Action :
```
name: ALL_deploy-production
on:
  push:
    branches: [ master ]
    commit-ignore:
      - 🚀
      - [bot]
      - Update from Shopify
  workflow_dispatch:
jobs:
  build:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: actions/checkout@v2
      - name: Merge
        run: |

          regionMasterBranches=("au-master" "us-master")

          git remote add github https://$GITHUB_ACTOR:$GITHUB_TOKEN@github.com/$GITHUB_REPOSITORY.git
          git config user.email email@email.com
          git config user.name My-Username
          git pull

          # Get all files and dirs except for templates, locales and config from master
          # get all file name in current directory, except templates, config, locales or . and stores them in an array 
          mapfile masterFiles <<< "$(find . -maxdepth 1 ! -name templates ! -name config ! -name locales ! -name . ! -name .git)"
          echo "MASTER FILES $masterFiles"
          # echo "MASTER FILES: $masterFiles"
          
          # Get the file names and statuses of the last merge commit and strip out the deleted files and format onto a single line
          deletedFiles="$(git log -m -1 --name-status --pretty="format:" | awk '/^D/ {print $2}' | tr '\n' ' ')"
          echo "DELETED FILES $deletedFiles"

          lastCommit="$(git log -m -1 --pretty="format:%H")"
          echo "LAST COMMIT - $lastCommit"

          for i in ${!regionMasterBranches[@]}; do
            regionMasterBranch=${regionMasterBranches[$i]}
            git checkout $regionMasterBranch
            git pull

            for file in "${masterFiles[@]}"; do
              file="$(tr -d '\n' <<< "$file")"
              git checkout master $file
            done

            if [[ $deletedFiles =~ [A-za-z] ]]; then
              echo "has deleted files"
              git rm -r $deletedFiles
            fi

            git add .
            git commit -m "Added master updates to $regionMasterBranch"
            git push github $regionMasterBranch
          done
```

Merge commit details:
<img width="1792" alt="image" src="https://github.com/community/community/assets/91449114/40749bf9-b180-44c9-8e63-48d44ab1fe7b">

This is what I am expecting the $deletedFiles variable to output. On my local, this outputs the below string. In the github action, this evaluates to an empty string
```
DELETED FILES snippets/account-page.liquid snippets/icon.cart.liquid
```

Has anyone experienced issues such as this in the past? I'd be greatful for any input and advice. Thankyou.

**Comment by EDMPhoenix** [comment, marked as answer]
Based on your script and the problem you described, it could be because of how the GitHub Checkout Action behaves. By default, when GitHub Actions checks out the code for your repository, it only retrieves a single commit and doesn't fetch the entire Git history.

When you execute this command: 

```bash
deletedFiles="$(git log -m -1 --name-status --pretty="format:" | awk '/^D/ {print $2}' | tr '\n' ' ')"
```

You are trying to find the deleted files from the last merge commit, but since GitHub Actions only fetches a single commit for efficiency and speed, it won't have the full Git log for your repository, explaining why your `deletedFiles` variable is empty.

You need to modify the `actions/checkout@v2` step to fetch the entire Git history so you can do your operations. Here's how to do that:

```yaml
- name: Checkout
  uses: actions/checkout@v2
  with:
    fetch-depth: 0
```

By setting `fetch-depth: 0`, the Checkout action will fetch all history for branches and tags, which should enable your script to work correctly.

Note: Fetching the entire Git history might slow down your workflow because it increases the amount of data that needs to be downloaded. If possible, you might want to find a more efficient way to detect deleted files that doesn't require the entire Git history. 

Let me know if this helps or if you need further clarification.

**Reply by JoshMealing** [reply]
Hey @EdmTechPhoenix, thank you so much for taking the time to reply and explain what the cause of my issue was. 

I made the update and tested and it is working perfectly. 

Whilst fetching the entire git history may not be the most efficient, I think for the time being this will be fine for my use case.

I will however look for additional ways of doing this. Did you know of any potential methods?
